### ISSUE TITLE
Problems setting up Notebook environment

### ISSUE THREAD

**Issue by Legsleg** [datalayer/jupyter-ui#113]
As part of a university project, we, a small group of students, have the task of integrating Jupyter Notebooks into a web page and building upon it. When we stumbled upon this project, it looked promising in easing the pain of the integration part since Jupyter does not really seem to be built for that kind of usage. But trying to utilise it, we ran into some problems. I figured this is the only place to ask you questions, so I am posting here. I hope you can help us interpret some of the console errors we are given.

# Our Goal
The main goal, for now, is the integration of a single notebook into a web page (the capability of switching notebooks is supposed to be implemented manually)
We tried the Notebook component as well as the JupyterLabApp component. They are both only configured with a path to the required file. The Jupyter server is its own instance running locally but configured with the config file provided in the examples. Credentials are given in the index.html (like in the examples).
In the long term, we want to use JupyterHub, but jupyter-ui should only be used to display and edit individual notebooks. We plan to implement the steps for the interaction before that ourselves.

# Notebook component
Using the Notebook component looks like this:
![Screenshot 2023-11-30 142725](https://github.com/datalayer/jupyter-ui/assets/64226008/e70f6fc6-adf0-4ff1-a021-1eea70822e0e)
## What works
- The correct notebook is loaded
- Cells can be run and added
- Embeds quite nicely into the web page

## What does not work
- I did not find a way to save the notebook (changes are lost on reload)
- ipywidgets does not work (for error, see next section)

## Errors
Loading the page causes some errors, and I need help with interpreting. From my understanding, they both have something to do with ipywidgets.

This error is shown independently of actually trying to run a cell using ipywidgets:
![Screenshot 2023-11-30 141510](https://github.com/datalayer/jupyter-ui/assets/64226008/013b576d-050e-414b-8841-5cd284f70359)

This error is only shown if a cell using ipywidgets is run:
![Screenshot 2023-11-30 142636](https://github.com/datalayer/jupyter-ui/assets/64226008/3df411d7-8193-4f59-b182-98444dc5322d)

# JupyterLabApp component
Using the JupyterLabApp component looks like this:
![Screenshot 2023-11-30 144641](https://github.com/datalayer/jupyter-ui/assets/64226008/16a495a4-a145-4e08-a3ed-4e30816efe27)

## What does work
- The correct notebook is loaded
- Cells can be run and added
- Notebooks can be saved

## What does not work
- Less focus on the individual notebook (lab layout with navigation capabilities is shown)
- ipywidgets does not throw an error, but the actual slider is not shown (as opposed to using the regular JupyterLab interface, where it is shown correctly)
- component sometimes takes over the whole screen (scrolling out of it seems to be buggy, probably fixable by manually limiting size)

# Conclusion
I spent quite some time getting the setup to where we are now, but I am now at a point where I need help. I hope you can provide some pointers on what could cause the problems above, where changes occurred in the current implementation relative to the things documented on the internet (for example, gifs showing the examples seem to not be up-to-date) and whether the required functionality is currently working.

**Comment by Legsleg** [comment]
Thank you for your lightning-fast response!

# Notebook Component
> There is a redux action for that, see this example

Thank you for sending the example! Unfortunately, I still could not quite piece it together. What do I have to pass into the uid field?

> Which version are you using?

We are currently using version 0.8.0

# JupyterLabApp Component
> Not sure to understand, could you rephrase/expand?

The wording, indeed, could have been clearer. The question is whether there is an option to deactivate some visual parts to allow for more focus on the single notebook, even when using the JupyterLabApp component. I could not find such an option (and also do not expect it; it just would be nice for our use case).

> from the top of my head, the ipywidget extension should be added (I think it it is not)

Could you further elaborate on that? Judging by the fact I don't know what you mean by "adding" it, you might be on the right track.

> can you record a animated GIF to see what issue you are facing?

Here is a gif of what I mean. As you can see, the component takes up the whole screen when loaded completely, and it is impossible to scroll upwards. Moving the mouse to the top left makes the navbar pop back into view - after that, scrolling back down to see the whole component isn't possible.
![Animation](https://github.com/datalayer/jupyter-ui/assets/64226008/d88132b4-dd18-4d38-a702-974c9dd2517e)
This issue can be mitigated (as I expected) by setting the height to be smaller than the available vertical space.

While working with the JupyterLabApp component over the past few days, I also encountered the following problems:
- The path to themes, as adopted from the examples, does not seem to work in our standard setup, leading to an error when mounting the component (as seen in the gif). Is there some way to deactivate that feature since it is irrelevant to us for now?
- Popups that normally would cover the whole screen are shown below the actual component. (as seen in the gif)
- The component is blank when repeatedly mounting it without reloading the page, for example, when switching between sub-pages using the React Router. The console still outputs the information shown when the component is mounted, but no content is displayed. This notably does not happen with the Notebook component (it works as expected even on repeated remounts).

**Comment by fcollonval** [comment]
Hey @Legsleg 

### Notebook Component

> Thank you for sending the example! Unfortunately, I still could not quite piece it together. What do I have to pass into the uid field?

If you are using the `Notebook` component, the uid pass to the toolbar will be set from the one specified as `Notebook` props.

https://github.com/datalayer/jupyter-ui/blob/4a68e971f642081a909872987990313273f9e341/packages/react/src/examples/Notebook.tsx#L15-L27

Internally jupyter-ui can handle multiple notebooks. The uid is the key used to discriminate between them.

**Comment by fcollonval** [comment]
### JupyterLabApp component

> Scrolling issue

It seems to me that to be able to scroll up again, you will need to play with the CSS of the all page to handle the [overflow](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow).

> Popups that normally would cover the whole screen are shown below the actual component.

I would also look at the CSS rules to fix that one. Feel free to open a PR if you find a good fix.

**Comment by fcollonval** [comment]
Hey @Legsleg,

When executing the _Notebook_ example locally, I see that error. But it does not prevent me to execute the cell that will then display the interactive widget correctly.

![Image](https://github.com/datalayer/jupyter-ui/assets/8435071/1e3c37be-24f9-4277-a73e-766f74cdfea1)

Do you observe the same behavior in your case?

---

@echarles I tested the `IPyWidgets` example with a notebook storing the widgets state. But that results also in a _Comm not found_ error that breaks rendering of the widgets at first load.

**Comment by echarles** [comment]
> @echarles I tested the IPyWidgets example with a notebook storing the widgets state. But that results also in a Comm not found error that breaks rendering of the widgets at first load.

Thx for testing. That is something we will need to look at. I have opened https://github.com/datalayer/jupyter-ui/issues/126 for this.

**Comment by Legsleg** [comment]
@fcollonval I just got back to working on the relevant components the past few days and also oberve the behaviour shown in your gif in my setup. Still, when the cell is executed manually, although it shows the widget correctly, an exception is logged in the console ("Exception opening new comm", as can also be seen in your gif).

Since my initial questions are solved, I will close this issue. Thank you for helping me get started and for creating this helpful package in the first place!

**Comment by Legsleg** [comment]
@fcollonval I just got back to working on the relevant components the past few days and also oberve the behaviour shown in your gif in my setup. Still, when the cell is executed manually, although it shows the widget correctly, an exception is logged in the console ("Exception opening new comm", as can also be seen in your gif).

Since my initial questions are solved, I will close this issue. Thank you for helping me get started and for creating this helpful package in the first place!
